lib/api/readable.js

```diff
@@ -174,26 +273,34 @@ class BodyReadable extends Readable {
     }
 
     return await new Promise((resolve, reject) => {
-      if (this[kContentLength] > limit || this[kBytesRead] > limit) {
+      if (
+        (this[kContentLength] && (this[kContentLength] > limit)) ||
+        this[kBytesRead] > limit
+      ) {
         this.destroy(new AbortError())
       }
 
-      const onAbort = () => {
-        this.destroy(signal.reason ?? new AbortError())
+      if (signal) {
+        const onAbort = () => {
+          this.destroy(signal.reason ?? new AbortError())
+        }
+        signal.addEventListener('abort', onAbort)
+        this
+          .on('close', function () {
+            signal.removeEventListener('abort', onAbort)
+            if (signal.aborted) {
+              reject(signal.reason ?? new AbortError())
+            } else {
+              resolve(null)
+            }
+          })
+      } else {
+        this.on('close', resolve)
       }
```

Comment:
reorderering so that we only do signal specific stuff like creating onAbort if needed